pkg/providers/vsphere/upgrade/virtualmachine/vm_schema_upgrade.go

```diff
@@ -397,39 +406,358 @@ func reconcileSCSIController(
 
 		case *vimtypes.VirtualDisk:
-			reconcileVirtualDisk(ctx, vm, d)
+			diskDevices = append(diskDevices, d)
 
 		case *vimtypes.VirtualCdrom:
-			reconcileVirtualCDROM(ctx, vm, d)
+			cdromDevices = append(cdromDevices, d)
+		}
+	}
+
+	reconcileVirtualCDROMs(ctx, k8sClient, vm, ctrlKeyToCtrlStatus, cdromDevices)
+	reconcileVirtualDisks(ctx, k8sClient, vm, ctrlKeyToCtrlStatus, diskDevices)
+}
+
+// reconcileVirtualDisks reconciles the VM's disk devices during
+// schema upgrade. It maps disk devices from the vSphere VM
+// configuration to the VirtualMachine spec volumes, updating the
+// unit number, controller type, and controller bus number for each
+// PVC volume based on the disk UUID and controller information.
+func reconcileVirtualDisks(
+	ctx context.Context,
+	k8sClient ctrlclient.Client,
+	vm *vmopv1.VirtualMachine,
+	ctrlKeyToCtrlStatus map[int32]vmopv1.VirtualControllerStatus,
+	diskDevices []*vimtypes.VirtualDisk) {
+
+	logger := pkglog.FromContextOrDefault(ctx)
+	logger.V(4).Info("Reconciling schema upgrade for VM disks")
+
+	if vm.Status.BiosUUID == "" {
+		logger.V(4).Info(
+			"Skipping PVC volume backfilling, BiosUUID not available")
+		return
+	}
+
+	diskUUIDToDiskInfo := buildDiskUUIDToDiskInfoMap(diskDevices)
+	volumeNameToDiskUUID, err := buildVolumeNameToDiskUUIDMap(
+		ctx, k8sClient, vm)
+	if err != nil {
+		return
+	}
+
+	for i := range vm.Spec.Volumes {
+		vol := &vm.Spec.Volumes[i]
+
+		if vol.PersistentVolumeClaim == nil {
+			continue
 		}
+
+		reconcilePVCVolumePlacement(
+			ctx,
+			vol,
+			volumeNameToDiskUUID,
+			diskUUIDToDiskInfo,
+			ctrlKeyToCtrlStatus)
 	}
 }
 
-func reconcileVirtualDisk(
+// buildDiskUUIDToDiskInfoMap builds a map of disk UUID to disk
+// device info by extracting from vSphere VM hardware.
+func buildDiskUUIDToDiskInfoMap(
+	diskDevices []*vimtypes.VirtualDisk) map[string]pkgutil.VirtualDiskInfo {
+
+	diskUUIDToDiskInfo := make(map[string]pkgutil.VirtualDiskInfo)
+	for _, disk := range diskDevices {
+		vdi := pkgutil.GetVirtualDiskInfo(disk)
+		if vdi.UUID != "" {
+			diskUUIDToDiskInfo[vdi.UUID] = vdi
+		}
+	}
+	return diskUUIDToDiskInfo
+}
+
+// buildVolumeNameToDiskUUIDMap builds a map of volume name to
+// disk UUID by querying CnsNodeVmAttachment objects. This is the
+// source of truth for PVC volume attachments.
+func buildVolumeNameToDiskUUIDMap(
 	ctx context.Context,
-	_ *vmopv1.VirtualMachine,
-	_ *vimtypes.VirtualDisk) {
+	k8sClient ctrlclient.Client,
+	vm *vmopv1.VirtualMachine) (map[string]string, error) {
 
 	logger := pkglog.FromContextOrDefault(ctx)
-	logger.V(4).Info("Reconciling schema upgrade for VM disk")
+	volumeNameToDiskUUID := make(map[string]string)
+	attachments, err := pkgutil.GetAttachmentsForVM(ctx, k8sClient, vm)
+	if err != nil {
+		logger.V(4).Info(
+			"Failed to get CnsNodeVmAttachments", "error", err)
+		return nil, err
+	}
+
+	for _, attachment := range attachments {
+		if attachment.Status.Attached {
```

Comment:
JW - If it gets attached _just_ after the schema upgrade, what happens?

Reply:
Currently, all CD-ROMs and Volume specs are best-effort backfilling. We are not returning anything. So it will not get backfilled. 

The VM CR will add the annotations to trigger an update webhook, and the validation webhook will log the error. 
Any new updates will also need to fix the existing issue with vm.spec.volume.pvc has incomplete placement information. 

The user needs to manually fill it out based on the vm.status.hardware.controller information, or just delete and create a new one.